pkg/api/metrics.go

```diff
@@ -16,251 +16,288 @@ import (
+				QueryType:     query.Get("queryType").MustString(""),
+				JSON:          modelJSON,
+			},
+		})
 	}
 
-	statusCode := http.StatusOK
-	for _, res := range resp.Results {
-		if res.Error != nil {
-			res.ErrorString = res.Error.Error()
-			resp.Message = res.ErrorString
-			statusCode = http.StatusBadRequest
+	if !req.hasExpression {
+		if len(datasources) > 1 {
+			// We do not (yet) support mixed query type
+			return nil, NewErrBadQuery("all queries must use the same datasource")
 		}
 	}
 
-	return response.JSON(statusCode, &resp)
+	return req, nil
 }
 
-func (hs *HTTPServer) createRequest(ctx context.Context, ds *models.DataSource, query legacydata.DataQuery) (*backend.QueryDataRequest, error) {
-	instanceSettings, err := adapters.ModelToInstanceSettings(ds, hs.decryptSecureJsonDataFn())
-	if err != nil {
-		return nil, err
+func (hs *HTTPServer) getDataSourceFromQuery(user *models.SignedInUser, skipCache bool, query *simplejson.Json, history map[string]*models.DataSource) (*models.DataSource, error) {
+	var err error
+	uid := query.Get("datasource").Get("uid").MustString()
+
+	// before 8.3 special types could be sent as datasource (expr)
+	if uid == "" {
+		uid = query.Get("datasource").MustString()
 	}
 
-	if query.Headers == nil {
-		query.Headers = make(map[string]string)
+	// check cache value
+	ds, ok := history[uid]
+	if ok {
+		return ds, nil
 	}
 
-	if hs.OAuthTokenService.IsOAuthPassThruEnabled(ds) {
-		if token := hs.OAuthTokenService.GetCurrentOAuthToken(ctx, query.User); token != nil {
-			delete(query.Headers, "Authorization")
-			query.Headers["Authorization"] = fmt.Sprintf("%s %s", token.Type(), token.AccessToken)
-		}
+	if expr.IsDataSource(uid) {
+		return expr.DataSourceModel(), nil
 	}
 
-	req := &backend.QueryDataRequest{
-		PluginContext: backend.PluginContext{
-			OrgID:                      ds.OrgId,
-			PluginID:                   ds.Type,
-			User:                       adapters.BackendUserFromSignedInUser(query.User),
-			DataSourceInstanceSettings: instanceSettings,
-		},
-		Queries: []backend.DataQuery{},
-		Headers: query.Headers,
+	if uid == grafanads.DatasourceUID {
+		return grafanads.DataSourceModel(user.OrgId), nil
 	}
 
-	for _, q := range query.Queries {
-		modelJSON, err := q.Model.MarshalJSON()
+	if uid != "" {
+		ds, err = hs.DataSourceCache.GetDatasourceByUID(uid, user, skipCache)
 		if err != nil {
 			return nil, err
 		}
-		req.Queries = append(req.Queries, backend.DataQuery{
-			RefID:         q.RefID,
-			Interval:      time.Duration(q.IntervalMS) * time.Millisecond,
-			MaxDataPoints: q.MaxDataPoints,
-			TimeRange: backend.TimeRange{
-				From: query.TimeRange.GetFromAsTimeUTC(),
-				To:   query.TimeRange.GetToAsTimeUTC(),
-			},
-			QueryType: q.QueryType,
-			JSON:      modelJSON,
-		})
+		return ds, nil
 	}
 
-	return req, nil
+	// Fallback to the datasourceId
+	id, err := query.Get("datasourceId").Int64()
```

Comment:
This change as is currently suffers from the same problem that was recently fixed here https://github.com/grafana/grafana/pull/41446

Reply:
Nicely spotted. Weird we don't get a conflict for that one. Maybe handled conflicts badly earlier. I'll have a look at this first thing Monday